osu.Server.Spectator/Hubs/MultiplayerHub.cs

```diff
@@ -702,36 +717,73 @@ private async Task updateRoomStateIfRequired(ServerMultiplayerRoom room)
                 case MultiplayerRoomState.Playing:
                     if (room.Users.All(u => u.State != MultiplayerUserState.Playing))
                     {
-                        foreach (var u in room.Users.Where(u => u.State == MultiplayerUserState.FinishedPlay))
-                            await changeAndBroadcastUserState(room, u, MultiplayerUserState.Results);
+                        var playingUsers = room.Users.Where(u => u.State == MultiplayerUserState.FinishedPlay).ToArray();
+
+                        await changeAndBroadcastAllUserStates(room, playingUsers, MultiplayerUserState.Results, client);
 
                         await changeRoomState(room, MultiplayerRoomState.Open);
                         await Clients.Group(GetGroupId(room.RoomID)).ResultsReady();
 
-                        await room.Queue.FinishCurrentItem();
+                        await room.Queue.FinishCurrentItem(client);
                     }
 
                     break;
             }
         }
 
-        private async Task changeAndBroadcastUserState(ServerMultiplayerRoom room, MultiplayerRoomUser user, MultiplayerUserState state)
+        /// <summary>
+        /// Changes the states of all users in the given array.
+        /// </summary>
+        /// <param name="room">The room.</param>
+        /// <param name="users">The users whose states are to be changed.</param>
+        /// <param name="state">The new user state.</param>
+        /// <param name="existingClient">Any existing <see cref="MultiplayerClientState"/> usage.
+        /// Must not be <c>null</c> if a user has been retrieved via <see cref="StatefulUserHub{TClient,TUserState}.UserStates"/> or
+        /// <see cref="StatefulUserHub{TClient,TUserState}.GetOrCreateLocalUserState"/> prior to this call.</param>
+        private async Task changeAndBroadcastAllUserStates(ServerMultiplayerRoom room, MultiplayerRoomUser[] users, MultiplayerUserState state, MultiplayerClientState? existingClient)
         {
+            foreach (var u in users)
+            {
+                if (u.UserID == existingClient?.UserId)
+                    await changeAndBroadcastUserState(room, u, existingClient, state);
+                else
+                {
+                    using (var userUsage = await UserStates.GetForUse(u.UserID))
+                    {
+                        if (userUsage.Item != null)
+                            await changeAndBroadcastUserState(room, u, userUsage.Item, state);
+                    }
+                }
+            }
+        }
```

Comment:
```diff
diff --git a/osu.Server.Spectator/Hubs/MultiplayerHub.cs b/osu.Server.Spectator/Hubs/MultiplayerHub.cs
index c3fc519..1404816 100644
--- a/osu.Server.Spectator/Hubs/MultiplayerHub.cs
+++ b/osu.Server.Spectator/Hubs/MultiplayerHub.cs
@@ -737,15 +737,15 @@ namespace osu.Server.Spectator.Hubs
         /// <param name="room">The room.</param>
         /// <param name="users">The users whose states are to be changed.</param>
         /// <param name="state">The new user state.</param>
-        /// <param name="existingClient">Any existing <see cref="MultiplayerClientState"/> usage.
+        /// <param name="currentUser">The <see cref="MultiplayerClientState"/> for the current user context.
         /// Must not be <c>null</c> if a user has been retrieved via <see cref="StatefulUserHub{TClient,TUserState}.UserStates"/> or
         /// <see cref="StatefulUserHub{TClient,TUserState}.GetOrCreateLocalUserState"/> prior to this call.</param>
-        private async Task changeAndBroadcastAllUserStates(ServerMultiplayerRoom room, MultiplayerRoomUser[] users, MultiplayerUserState state, MultiplayerClientState? existingClient)
+        private async Task changeAndBroadcastAllUserStates(ServerMultiplayerRoom room, MultiplayerRoomUser[] users, MultiplayerUserState state, MultiplayerClientState? currentUser)
         {
             foreach (var u in users)
             {
-                if (u.UserID == existingClient?.UserId)
-                    await changeAndBroadcastUserState(room, u, existingClient, state);
+                if (u.UserID == currentUser?.UserId)
+                    await changeAndBroadcastUserState(room, u, currentUser, state);
                 else
                 {
                     using (var userUsage = await UserStates.GetForUse(u.UserID))

```

I think calling this the "current" user or "local" user reads better than "existing". I'd update all the other methods which have this parameter to match.